### ISSUE TITLE
BUG: Temp file with list of selected file names has missing entries

### ISSUE THREAD

**Issue by fitdev** [FarGroup/FarManager#240]
I use `!@FQ!` syntax in the User Command (F2 menu) to pass full path to a text file that _contains full paths of all selected files_ to a third party tool.

The problem is that often the resulting temp file that FAR creates has empty lines (corresponding to missing files, that even though they were selected in the panel, were not included in the list).

```
"D:\Dir\file 1.avi"


"D:\Dir\file 4.avi"
"D:\Dir\file 5.avi"
```

Note the 2 and 3 are missing, even though corresponding files have been selected.

I am using FAR 3.0.0.5590 x64 on Windows 10

**Comment by alabuzhev** [comment]
I cannot reproduce this issue.
Does it happen with only particular files/paths or any?
Do [previous versions](https://farmanager.com/nightly.php) work as expected?

**Comment by fitdev** [comment]
Thanks for looking into this. To answer your questions:
1. No, it seems to happen with all kinds of paths and there does not seem to be a pattern. Both ASCII-only paths and paths with Unicode characters are affected.
2. The issue has been around for at least 4 months or so (I update to nightly every couple of weeks), so it's not specific to the particular build I am using.
3. Here's the sample command:
```bat
for /F "usebackq tokens=*" %a in ("!@FQ!") do ffmpeg.exe -i %a -max_muxing_queue_size 1024 -c:v libx264 -crf !?CRF [15-35] (higher number = worse quality, rec 25)?25! -preset medium -c:a !?Audio (aac|copy)?copy! -map 0:0 -map 0:1? -map 0:2? -map 0:3? -map 0:4? -map 0:5? -map 0:6? -map 0:7? -threads !?Threads (0 - default)?4! "!?Directory (or empty)?!%~na.x264.!?Container (mp4|mkv)?mkv!"
```

**Comment by alabuzhev** [comment]
When it happens next time, could you please attach the generated list file?
Also, it would be really helpful if we find out when exactly the issue was introduced. Could you please try to narrow it down? With binary search it should not take too many attempts.
For example, take one that definitely works, say, 5500.
Check the middle between 5500 and 5590, which is 5545.
If it doesn't work, check the middle between 5545 and 5590, which is 5567, otherwise check the middle between 5500 and 5545, which is 5522 and so on.

**Comment by fitdev** [comment]
I will try to provide more info the next time it happens, but the problem is that the behavior is inconsistent. Sometimes the list has missing entries, and sometimes it does not, even if the selected files are the same.

Whenever it does not work, I would manually check (after the command has finished executing) which files were processed and which ones were not. I would then unselect the processed files, and leave the unprocessed files selected, re-running the command, and this time it would process more (if not all) the files. Though sometimes some files would be omitted again even on the second run, requiring me to repeat the procedure again - each time more and more files would be processed.

By "files would be processed" I mean that their paths are written to the file list file by FAR.

**Comment by fitdev** [comment]
Also, to clarify: the generated temp list file has _empty lines_ in place of quoted file paths. That is the lines are there - they are simply empty, without any characters.

**Comment by fitdev** [comment]
Update:

I thought it was fixed, as I have not encountered this issue for a while. But today, I encountered it again: the temp file with file paths had only 4 file paths out of 12. In place of the missing quoted paths there are just empty lines: ```\0d\0a```s.

I am now using FAR 3.0.5700 x64 on Windows 10 1803.

Some observations:

- The paths do contain non-ASCII characters
- If I repeat the procedure (i.e. select the files again in FAR, and invoke a command), this time _more_ files will be included in the temp file list (seemingly chosen at random).
- The files themselves are not special - neither hidden nor have any weird attributes.

**Comment by alabuzhev** [comment]
@fitdev could you please give a few examples of these paths with non-ASCII characters?
Also, what is your OEM codepage? (can be checked via Shift+F8 in viewer/editor, System section in the menu).

**Comment by fitdev** [comment]
The sample paths are (lines taken directly from `FAR18F3.tmp`):

```
"D:\Downloads\Temp\_40m2\03 Иииии Ддддддд 1997.avi"
"D:\Downloads\Temp\_40m2\08 Иииииии № 37 Ппп Вввввв... 1998.avi"
```

The codepage for the `FAR18F3.tmp` file generated by FAR is reported as UTF-8 (65001).

The strange part is that the omitted paths seem to be totally random. For example, when I attempted to run my F2 command (which uses FAR's temp file list function) today to check the codepage of the temp file as you requested, in that temp file _only 2 paths_ were missing, and the rest 20+ were there. Whereas yesterday, on the _same set of files_, _most paths were missing_, so I had to run my F2 command in batches basically, - every time checking which files were processed (i.e. whose paths were included in the temp file list), unselecting them, and then running the command again on the remaining files - I did this 3 or 4 times until all files have been eventually processed (included in the temp file list).

In other words, the 2 sample paths above will sometimes work, and sometimes will fail. I do not remember for sure if paths that contain only ASCII characters would also fail sometimes, but I think it is usually the case that only paths with non-ASCII characters fail sometimes.

**Comment by alabuzhev** [comment]
> The codepage for the `FAR18F3.tmp` file generated by FAR is reported as UTF-8 (65001)

That's strange. If you're using `!@FQ!` as mentioned earlier, it shouldn't be UTF-8.
Or maybe you have "Beta: Use Unicode UTF-8 for worldwide language support" checked in system locale settings?

**Comment by alabuzhev** [comment]
I think I found it.
5732.

**Comment by fitdev** [comment]
> Or maybe you have "Beta: Use Unicode UTF-8 for worldwide language support" checked in system locale settings?

Yes, I do have that checked.

I will try the new build ASAP and let you know.

**Comment by fitdev** [comment]
Sorry for the delayed response. I have just gotten around to testing it now.

Now the temp file does have all file paths every time, so this is working correctly now. However, the generated tmp file is being prefixed with UTF header. As a result my commands no longer work, i.e. first line (path) is not being understood properly.

I actually had an issue with this behavior when using F4 editor to create new files.

To address both of these issues, perhaps you can introduce a system-wide setting that will determine whether the generated, as well as editor-created UTF8 files should have BOM header or not. Thank you.

**Comment by w17** [comment]
For editor -- try to set **far:config** Editor.AddUnicodeBOM = false.

**Comment by fitdev** [comment]
@w17 Thanks for the tip. But I am not sure I understood how / where do I access that setting?

**Comment by w17** [comment]
far:config <Enter>
in far command line
select mentioned parameter and press Enter to change value

**Comment by fitdev** [comment]
@w17 Thanks for your help! Got it! Wow I have been using FAR for 10+ years and never knew about `far:config`!

It works great for the Editor, but _not_ for the tmp file though (i.e. this issue).

**Comment by alabuzhev** [comment]
@fitdev I guess we never need BOM in list files anyway, so I removed it unconditionally for now.
5470 / c4fe9ab36

**Comment by fitdev** [comment]
Thank you for your fast fixes. I think this issue is now addressed.

**Comment by fitdev** [comment]
Hello! Having just updated to the latest stable build 5858, I see now that BOM is back in the list files, causing for instance ffmpeg to miss the first file in my scripts. Yet, the option in far:config for BOM in the editor is off (i.e. it should avoid BOM). Is there a way to not write BOM for list files?

**Comment by alabuzhev** [comment]
@fitdev yeah, my bad. Fixed (again) in 5885 / 2dc6c67.

**Comment by fitdev** [comment]
Thank you for a quick fix! Is 5885 safe to use though? In the last year after a few issues I was reluctant to update to nightly builds?

**Comment by alabuzhev** [comment]
> Is 5885 safe to use though?

There's only one way to find out.
